WordPress/Classes/Services/CommentServiceRemoteCoreRESTAPI.swift

```diff
@@ -0,0 +1,323 @@
+private extension RemoteComment {
+    convenience init(comment: CommentWithEditContext) {
+        self.init()
+
+        self.commentID = NSNumber(value: comment.id)
+        self.authorID = NSNumber(value: comment.author)
+        self.author = comment.authorName
+        self.authorEmail = comment.authorEmail
+        self.authorUrl = comment.authorUrl
+        self.authorAvatarURL = comment.authorAvatarUrls.avatarURL()?.absoluteString
+        self.authorIP = comment.authorIp
+        self.content = comment.content.raw
+        self.rawContent = comment.content.raw
+        self.date = comment.dateGmt
+        self.link = comment.link
+        self.parentID = NSNumber(value: comment.parent)
+        self.postID = NSNumber(value: comment.post)
+
+        self.status = comment.status.commentStatusType?.description
+        self.type = comment.commentType.type
+
+        // The following properties are not available in .org REST API.
+        self.postTitle = nil
+        self.isLiked = false
+        self.likeCount = 0
+        self.canModerate = false
+    }
+
+    /// When `RemoteComment` is created by the app to send to API, some properties are set to default values (like
+    /// empty string, 0). We should treat those values as `nil`, because wordpress-rs emits nil values in request body,
+    /// but keeps them if they are "" or 0.
+    ///
+    /// We'll treat each property case by case, but in general it's safe to treat "" or 0 as nil. For example, there is
+    /// no post whose id is 0, and there is no email address is an empty string.
+    func resetDefaultValuesToNil() {
+        if self.commentID?.intValue == 0 { self.commentID = nil }
+        if self.authorID?.intValue == 0 { self.authorID = nil }
+        if self.author?.isEmpty == true { self.author = nil }
+        if self.authorEmail?.isEmpty == true { self.authorEmail = nil }
+        if self.authorUrl?.isEmpty == true { self.authorUrl = nil }
+        if self.authorAvatarURL?.isEmpty == true { self.authorAvatarURL = nil }
+        if self.authorIP?.isEmpty == true { self.authorIP = nil }
+        if self.content?.isEmpty == true { self.content = nil }
+        if self.rawContent?.isEmpty == true { self.rawContent = nil }
+        if self.link?.isEmpty == true { self.link = nil }
+        if self.parentID?.intValue == 0 { self.parentID = nil }
+        if self.postID?.intValue == 0 { self.postID = nil }
+        if self.postTitle?.isEmpty == true { self.postTitle = nil }
+        if self.status?.isEmpty == true { self.status = nil }
+        if self.type?.isEmpty == true { self.type = nil }
+
+        // The following properties are not updated:
+        // - likeCount
+    }
+
+    /// The `status: String` property value does not match the `CommentStatus` values. This variable provides a safe
+    /// way to cast the `status` property to a `CommentStatus` instance.
+    var commentStatus: CommentStatus? {
+        if let status = CommentStatusType.typeForStatus(status) {
```

Comment:
It seems we have a fallthrough `custom` variant for `CommentStatus`, so how could this return `nil`?

Reply:
Because the meaning of `CommentStatusType`'s string value is _not_ the same as `CommentStatus`.

`CommentStatus`'s string comes from the API; it can hold any random string. `CommentStatusType` is an arbitrary type defined by the app, with [only five possible string values](https://github.com/wordpress-mobile/WordPress-iOS/blob/7a1d4304840a2a65254fe0d76f8361ba74eb1ae0/Sources/WordPressData/Swift/Comment%2BCoreDataClass.swift#L185-L191), which are hard-coded strings.

The `status` variable here should be one of those five possible values in `CommentStatusType`. Since they are hard-coded values, we can assume they are the values that the REST API accepts and use them to create `CommentStatus`.

Reply:
For example, REST API returns `"approved"` for approved comment type. But `CommentStatusType.approve`'s string value is `"approve"`.

If we just blindly use `CommentStatusType`'s string to create `CommentStatus`, we'll get `CommentStatus.custom("approve")`, which the REST API does not understand. That causes a bug in the app, where you try to approve a comment, but the REST API won't do it, because it expects "approved", not "approve".